### PR TITLE
Bootcamp completed.v1(Yue Yang)

### PR DESCRIPTION
**Pull request by yyggzsCN**

### CODE REVIEW

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 1]
```diff
@@ -70,11 +76,52 @@ def run(self,
 
         # Do something based on the report and the state of this class...
 
+        # get the distance bewteen pad and drone
+        def get_distance(landing_pad_location: "location.Location",
```
Comment (AaronWang04) [inline modules/bootcamp/decision_waypoint_landing_pads.py:80]: declearing function within another function means it is called everytime the main function is ran. since the inner function is not dependent on the state of the outer function, and you don't need to encapsulate anything in particular, it is a good idea to declear within the class rather than the function

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 2]
```diff
@@ -70,11 +76,52 @@ def run(self,
...
+            result = x ** 2 + y ** 2
+            return result
+        
+        # get the radius of the drone. 
+        #def get_position_radius(dron_position: "location.Location") -> float:
+        #    return dron_position.location_x ** 2 + dron_position.location_y ** 2
+        
+        # Program started and go to the way point
+        if self.travaling == False and report.status == drone_status.DroneStatus.HALTED and not self.travaling:
+                command = self.commands[0]
```
Comment (AaronWang04) [inline modules/bootcamp/decision_waypoint_landing_pads.py:94]: double indentation

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 3]
```diff
@@ -70,11 +76,52 @@ def run(self,
...
+                self.travaling = True
+        
+        #drone is at the waypoint and decide which pad is the closest.
+        elif self.go_to_closest_pad == False and self.travaling == True and report.status == drone_status.DroneStatus.HALTED:
+            
+            # get the closest pad position.
+            min_distance = 100000000000000
```
Comment (AaronWang04) [inline modules/bootcamp/decision_waypoint_landing_pads.py:101]: you don't need to do this, you can set the first landing pad's distance as the "min_distance"

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 4]
```diff
@@ -70,11 +76,52 @@ def run(self,
...
-        raise NotImplementedError
+        #raise NotImplementedError
 
         # ============
         # ↑ BOOTCAMPERS MODIFY ABOVE THIS COMMENT ↑
         # ============
 
         return command
+    
+
```
Comment (AaronWang04) [inline modules/bootcamp/decision_waypoint_landing_pads.py:127]: empty lines at bottom of file does not follow style guide

File: modules/bootcamp/detect_landing_pad.py [hunk 5]
```diff
@@ -86,32 +87,37 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
         # Hint: .shape gets the dimensions of the numpy array
         # for i in range(0, ...):
             # Create BoundingBox object and append to list
             # result, box = ...
-
-        # Remove this when done
-        raise NotImplementedError
+        for i in range(boxes_cpu.shape[0]):
+            results, bbox = bounding_box.BoundingBox.create(boxes_cpu[i])
+            if results:
```
Comment (AaronWang04) [inline modules/bootcamp/detect_landing_pad.py:117]: note that if any of the results is false, we _should_ discard the whole the entire datapoint. (standard practice in stats)

try handling with an else statement

File: modules/bootcamp/detect_landing_pad.py [hunk 6]
```diff
@@ -86,32 +87,37 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
+        predictions = self.__model.predict(source=image, conf=0.8, device=self.__DEVICE, verbose=False)
         # Get the Result object
-        prediction = ...
-
+        
+        for prediction in predictions:
         # Plot the annotated image from the Result object
         # Include the confidence value
-        image_annotated = ...
+            image_annotated = prediction.plot(conf=True)
```
Comment (AaronWang04) [inline modules/bootcamp/detect_landing_pad.py:96]: just a note, you are essentially plotting the last image only since the jpgs all use the same default name

**Review by AaronWang04 — COMMENTED** [review]
good progress on the bootcamp! i made some comments but if any were unclear you can add your own reply and then ping me on discord

File: modules/bootcamp/decision_waypoint_landing_pads.py (hunk 4 above)
Comment (yyggzsCN) [reply modules/bootcamp/decision_waypoint_landing_pads.py:127]: resolved

File: modules/bootcamp/decision_waypoint_landing_pads.py (hunk 3 above)
Comment (yyggzsCN) [reply modules/bootcamp/decision_waypoint_landing_pads.py:101]: resolved

File: modules/bootcamp/decision_waypoint_landing_pads.py (hunk 2 above)
Comment (yyggzsCN) [reply modules/bootcamp/decision_waypoint_landing_pads.py:94]: resolved

File: modules/bootcamp/decision_simple_waypoint.py [hunk 7]
```diff
@@ -67,11 +67,23 @@ def run(self,
         # ============
         # ↓ BOOTCAMPERS MODIFY BELOW THIS COMMENT ↓
         # ============
+        def get_position_radius(dron_position: "location.Location") -> float:
+            return dron_position.location_x ** 2 + dron_position.location_y ** 2
```
[the remark this answers is not in the document]
Comment (yyggzsCN) [reply modules/bootcamp/decision_simple_waypoint.py:71]: resolved

File: modules/bootcamp/detect_landing_pad.py (hunk 5 above)
Comment (yyggzsCN) [reply modules/bootcamp/detect_landing_pad.py:117]: Is discarding the entire datapoint means set the bouding_boxes to default? 
ie. 
bonding_boxes = []
break

File: modules/bootcamp/decision_waypoint_landing_pads.py (hunk 1 above)
Comment (yyggzsCN) [reply modules/bootcamp/decision_waypoint_landing_pads.py:80]: resolved

File: modules/bootcamp/detect_landing_pad.py (hunk 6 above)
Comment (yyggzsCN) [reply modules/bootcamp/detect_landing_pad.py]: Is not that it only plot only one image every time the run() function is called?

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 8]
```diff
@@ -67,14 +85,39 @@ def run(self,
...
         # Remove this when done
-        raise NotImplementedError
+        #raise NotImplementedError
 
         # ============
         # ↑ BOOTCAMPERS MODIFY ABOVE THIS COMMENT ↑
         # ============
 
-        return command
+        return command
```
Comment (AaronWang04) [inline modules/bootcamp/decision_waypoint_landing_pads.py:123]: my bad i should've specified here. you should have a singular empty line with 0 space at end of the program. this never matters for python, but its a good convention set in programming. 

tldr it has smth to do with how unix terminates when reading the file. it ensures some low lvl unix tools function as intended.

nit pick comment :P

**Review by AaronWang04 — COMMENTED** [review]
some remaining comments, good progress!

**Review by AaronWang04 — APPROVED** [review]
